File: funtime/pylib/tictactoe_draw.py
"""tictactoe_draw.py - everything you can see in Tic-Tac-Toe, in Python."""

import math

from tictactoe_rules import GRID_SIZE, EMPTY, PLAYER, square_index

CELL_SIZE = 100
BOARD_MARGIN = 12

COLOR_INK = "#111111"
COLOR_PAPER = "#ffffff"
COLOR_FAINT = "#bbbbbb"


def board_pixel_size():
    """How big the square canvas must be."""
    return BOARD_MARGIN * 2 + GRID_SIZE * CELL_SIZE


def cell_left(column):
    """The pixel position of a square's left edge."""
    return BOARD_MARGIN + column * CELL_SIZE


def cell_top(row):
    """The pixel position of a square's top edge."""
    return BOARD_MARGIN + row * CELL_SIZE


def square_at_pixel(x, y):
    """Which square did the player click on? Returns (column, row) or None."""
    column = int((x - BOARD_MARGIN) // CELL_SIZE)
    row = int((y - BOARD_MARGIN) // CELL_SIZE)
    if not (0 <= column < GRID_SIZE and 0 <= row < GRID_SIZE):
        return None
    return (column, row)


def clear_canvas(ctx, width, height, color):
    """Paint the whole canvas one flat colour."""
    ctx.fillStyle = color
    ctx.fillRect(0, 0, width, height)


def draw_grid_lines(ctx):
    """The two vertical and two horizontal bars."""
    ctx.strokeStyle = COLOR_INK
    ctx.lineWidth = 6
    ctx.lineCap = "round"
    for line in range(1, GRID_SIZE):
        ctx.beginPath()
        ctx.moveTo(cell_left(line), BOARD_MARGIN + 6)
        ctx.lineTo(cell_left(line), BOARD_MARGIN + GRID_SIZE * CELL_SIZE - 6)
        ctx.stroke()
        ctx.beginPath()
        ctx.moveTo(BOARD_MARGIN + 6, cell_top(line))
        ctx.lineTo(BOARD_MARGIN + GRID_SIZE * CELL_SIZE - 6, cell_top(line))
        ctx.stroke()


def draw_mark(ctx, mark, column, row):
    """One X (two crossing lines) or one O (a circle)."""
    if mark == EMPTY:
        return

    x = cell_left(column)
    y = cell_top(row)
    pad = 24

    ctx.strokeStyle = COLOR_INK
    ctx.lineWidth = 10
    ctx.lineCap = "round"

    if mark == PLAYER:
        ctx.beginPath()
        ctx.moveTo(x + pad, y + pad)
        ctx.lineTo(x + CELL_SIZE - pad, y + CELL_SIZE - pad)
        ctx.stroke()
        ctx.beginPath()
        ctx.moveTo(x + CELL_SIZE - pad, y + pad)
        ctx.lineTo(x + pad, y + CELL_SIZE - pad)
        ctx.stroke()
    else:
        ctx.beginPath()
        ctx.arc(x + CELL_SIZE / 2, y + CELL_SIZE / 2, CELL_SIZE / 2 - pad, 0, math.pi * 2)
        ctx.stroke()


def draw_winning_line(ctx, line):
    """Strike through the three winning squares."""
    first, last = line[0], line[2]
    from_x = cell_left(first % GRID_SIZE) + CELL_SIZE / 2
    from_y = cell_top(first // GRID_SIZE) + CELL_SIZE / 2
    to_x = cell_left(last % GRID_SIZE) + CELL_SIZE / 2
    to_y = cell_top(last // GRID_SIZE) + CELL_SIZE / 2

    ctx.strokeStyle = COLOR_INK
    ctx.lineWidth = 8
    ctx.beginPath()
    ctx.moveTo(from_x, from_y)
    ctx.lineTo(to_x, to_y)
    ctx.stroke()


def draw_cursor(ctx, cursor):
    """The dashed box showing where the keyboard cursor is."""
    ctx.strokeStyle = COLOR_FAINT
    ctx.lineWidth = 4
    ctx.setLineDash([8, 6])
    ctx.strokeRect(cell_left(cursor["column"]) + 6, cell_top(cursor["row"]) + 6,
                   CELL_SIZE - 12, CELL_SIZE - 12)
    ctx.setLineDash([])


def draw_message(ctx, width, height, title, subtitle):
    """Big centred words across the board."""
    ctx.fillStyle = "rgba(255, 255, 255, 0.9)"
    ctx.fillRect(0, 0, width, height)
    ctx.fillStyle = COLOR_INK
    ctx.textAlign = "center"

    scale = width / 340
    title_size = max(14, round(32 * scale))
    subtitle_size = max(9, round(15 * scale))

    ctx.font = "bold " + str(title_size) + "px monospace"
    ctx.fillText(title, width / 2, height / 2 - 8)
    ctx.font = str(subtitle_size) + "px monospace"
    ctx.fillText(subtitle, width / 2, height / 2 + title_size * 0.8)


def render_game(ctx, state):
    """Draw one complete frame."""
    size = board_pixel_size()
    clear_canvas(ctx, size, size, COLOR_PAPER)
    draw_grid_lines(ctx)

    for row in range(GRID_SIZE):
        for column in range(GRID_SIZE):
            draw_mark(ctx, state["board"][square_index(column, row)], column, row)

    if not state["is_over"]:
        draw_cursor(ctx, state["cursor"])
    if state["line"]:
        draw_winning_line(ctx, state["line"])

    if state["is_over"]:
        if state["winner"] == PLAYER:
            title = "YOU WIN!"
        elif state["winner"] == "draw":
            title = "A DRAW"
        else:
            title = "COMPUTER WINS"
        draw_message(ctx, size, size, title, "Press R or tap New for another round")
    elif state["is_paused"]:
        draw_message(ctx, size, size, "PAUSED", "Press P or tap Play to carry on")

File: funtime/lib/tictactoe-draw.js
/* ============================================================
   tictactoe-draw.js — everything you can see in Tic-Tac-Toe
   ============================================================ */

const CELL_SIZE = 100;
const BOARD_MARGIN = 12;

const COLOR_INK = '#111111';
const COLOR_PAPER = '#ffffff';
const COLOR_FAINT = '#bbbbbb';

/** boardPixelSize — how big the square canvas must be. */
function boardPixelSize() {
    return BOARD_MARGIN * 2 + GRID_SIZE * CELL_SIZE;
}

/** cellLeft / cellTop — where one square sits on the canvas. */
function cellLeft(column) { return BOARD_MARGIN + column * CELL_SIZE; }
function cellTop(row) { return BOARD_MARGIN + row * CELL_SIZE; }

/**
 * squareAtPixel — which square did the player click on?
 * INPUT: x, y. OUTPUT: { column, row } or null.
 */
function squareAtPixel(x, y) {
    const column = Math.floor((x - BOARD_MARGIN) / CELL_SIZE);
    const row = Math.floor((y - BOARD_MARGIN) / CELL_SIZE);
    if (column < 0 || column >= GRID_SIZE || row < 0 || row >= GRID_SIZE) {
        return null;
    }
    return { column: column, row: row };
}

/** clearCanvas — paint the whole canvas one flat colour. */
function clearCanvas(ctx, width, height, color) {
    ctx.fillStyle = color;
    ctx.fillRect(0, 0, width, height);
}

/** drawGridLines — the two vertical and two horizontal bars. */
function drawGridLines(ctx) {
    ctx.strokeStyle = COLOR_INK;
    ctx.lineWidth = 6;
    ctx.lineCap = 'round';
    for (let line = 1; line < GRID_SIZE; line++) {
        ctx.beginPath();
        ctx.moveTo(cellLeft(line), BOARD_MARGIN + 6);
        ctx.lineTo(cellLeft(line), BOARD_MARGIN + GRID_SIZE * CELL_SIZE - 6);
        ctx.stroke();
        ctx.beginPath();
        ctx.moveTo(BOARD_MARGIN + 6, cellTop(line));
        ctx.lineTo(BOARD_MARGIN + GRID_SIZE * CELL_SIZE - 6, cellTop(line));
        ctx.stroke();
    }
}

/**
 * drawMark — one X or one O.
 * INPUT: ctx, mark, column, row. OUTPUT: nothing.
 * ALGORITHM: an X is two crossing lines; an O is a circle. Both are drawn
 *            inside the square with a comfortable margin.
 */
function drawMark(ctx, mark, column, row) {
    if (mark === EMPTY) { return; }

    const x = cellLeft(column);
    const y = cellTop(row);
    const pad = 24;

    ctx.strokeStyle = COLOR_INK;
    ctx.lineWidth = 10;
    ctx.lineCap = 'round';

    if (mark === PLAYER) {
        ctx.beginPath();
        ctx.moveTo(x + pad, y + pad);
        ctx.lineTo(x + CELL_SIZE - pad, y + CELL_SIZE - pad);
        ctx.stroke();
        ctx.beginPath();
        ctx.moveTo(x + CELL_SIZE - pad, y + pad);
        ctx.lineTo(x + pad, y + CELL_SIZE - pad);
        ctx.stroke();
    } else {
        ctx.beginPath();
        ctx.arc(x + CELL_SIZE / 2, y + CELL_SIZE / 2, CELL_SIZE / 2 - pad, 0, Math.PI * 2);
        ctx.stroke();
    }
}

/** drawWinningLine — strike through the three winning squares. */
function drawWinningLine(ctx, line) {
    const first = line[0];
    const last = line[2];
    const fromX = cellLeft(first % GRID_SIZE) + CELL_SIZE / 2;
    const fromY = cellTop(Math.floor(first / GRID_SIZE)) + CELL_SIZE / 2;
    const toX = cellLeft(last % GRID_SIZE) + CELL_SIZE / 2;
    const toY = cellTop(Math.floor(last / GRID_SIZE)) + CELL_SIZE / 2;

    ctx.strokeStyle = COLOR_INK;
    ctx.lineWidth = 8;
    ctx.beginPath();
    ctx.moveTo(fromX, fromY);
    ctx.lineTo(toX, toY);
    ctx.stroke();
}

/** drawCursor — the dashed box showing where the keyboard cursor is. */
function drawCursor(ctx, cursor) {
    ctx.strokeStyle = COLOR_FAINT;
    ctx.lineWidth = 4;
    ctx.setLineDash([8, 6]);
    ctx.strokeRect(cellLeft(cursor.column) + 6, cellTop(cursor.row) + 6, CELL_SIZE - 12, CELL_SIZE - 12);
    ctx.setLineDash([]);
}

/** drawMessage — big centred words across the board. */
function drawMessage(ctx, width, height, title, subtitle) {
    ctx.fillStyle = 'rgba(255, 255, 255, 0.9)';
    ctx.fillRect(0, 0, width, height);
    ctx.fillStyle = COLOR_INK;
    ctx.textAlign = 'center';

    const scale = width / 340;
    const titleSize = Math.max(14, Math.round(32 * scale));
    const subtitleSize = Math.max(9, Math.round(15 * scale));

    ctx.font = 'bold ' + titleSize + 'px monospace';
    ctx.fillText(title, width / 2, height / 2 - 8);
    ctx.font = subtitleSize + 'px monospace';
    ctx.fillText(subtitle, width / 2, height / 2 + titleSize * 0.8);
}

/** renderGame — draw one complete frame. */
function renderGame(ctx, state) {
    const size = boardPixelSize();
    clearCanvas(ctx, size, size, COLOR_PAPER);
    drawGridLines(ctx);

    for (let row = 0; row < GRID_SIZE; row++) {
        for (let column = 0; column < GRID_SIZE; column++) {
            drawMark(ctx, state.board[squareIndex(column, row)], column, row);
        }
    }

    if (!state.isOver) {
        drawCursor(ctx, state.cursor);
    }
    if (state.line) {
        drawWinningLine(ctx, state.line);
    }

    if (state.isOver) {
        const title = state.winner === PLAYER ? 'YOU WIN!'
            : (state.winner === 'draw' ? 'A DRAW' : 'COMPUTER WINS');
        drawMessage(ctx, size, size, title, 'Press R or tap New for another round');
    } else if (state.isPaused) {
        drawMessage(ctx, size, size, 'PAUSED', 'Press P or tap Play to carry on');
    }
}
